model/stacktrace.go

```diff
@@ -1,46 +1,58 @@
 package model
 
 import (
+	"errors"
+
 	pr "github.com/elastic/apm-server/processor"
 	"github.com/elastic/beats/libbeat/common"
 )
 
-type Stacktrace []*StacktraceFrame
+type Stacktrace struct {
+	Frames []*StacktraceFrame
+}
 
-func (st Stacktrace) Transform(config *pr.Config, service Service) []common.MapStr {
-	if config == nil || config.SmapMapper == nil {
-		return st.transform(config)
+func (st *Stacktrace) Decode(input interface{}) error {
+	raw, ok := input.([]interface{})
+	if raw == nil {
+		return nil
+	}
+	if !ok {
+		return errors.New("Invalid format for stacktrace")
 	}
-	return st.sourcemapAndTransform(config, service)
-}
 
-func (st Stacktrace) transform(config *pr.Config) []common.MapStr {
-	var frames []common.MapStr
-	for _, fr := range st {
-		frame := fr.Transform(config)
-		frames = append(frames, frame)
+	st.Frames = make([]*StacktraceFrame, len(raw))
+	for idx, fr := range raw {
+		frame := StacktraceFrame{}
+		err := frame.Decode(fr)
+		if err != nil {
+			return err
+		}
+		st.Frames[idx] = &frame
 	}
-	return frames
+	return nil
 }
 
-// source map algorithm:
-// apply source mapping frame by frame
-// if no source map could be found, set updated to false and set sourcemap error
-// otherwise use source map library for mapping and update
-// - filename: only if it was found
-// - function:
-//   * should be moved down one stack trace frame,
-//   * the function name of the first frame is set to <anonymous>
-//   * if one frame is not found in the source map, this frame is left out and
-//   the function name from the previous frame is used
-//   * if a mapping could be applied but no function name is found, the
-//   function name for the next frame is set to <unknown>
-// - colno
-// - lineno
-// - abs_path is set to the cleaned abs_path
-// - sourcmeap.updated is set to true
-func (st Stacktrace) sourcemapAndTransform(config *pr.Config, service Service) []common.MapStr {
-	frameCount := len(st)
+func (st *Stacktrace) Transform(config *pr.Config, service Service) []common.MapStr {
+	if st == nil {
+		return nil
+	}
+	// source map algorithm:
```

Comment:
why this comment is here? i think it fits better as `applySourcemap`

Reply:
because the comment describes the sourcemapping for the whole stacktrace and not just for a single frame.